geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/utils/StorageUtils.scala

```diff
@@ -43,10 +45,20 @@ object StorageUtils {
       fileType: FileType.FileType,
       name: String = UUID.randomUUID().toString.replaceAllLiterally("-", "")): Path = {
     val filename = s"$fileType$name.$extension"
-    val filenameWithLeaf = if (leaf) { s"${partition.split('/').last}_$filename" } else { filename }
+    val filenameWithLeaf = if (leaf) { s"${partition.split('/').last}$LeafSeparator$filename" } else { filename }
     new Path(baseDirectory(root, partition, leaf), filenameWithLeaf)
   }
 
+  /**
+   * Extract the 'leaf' part of a partition name from a file name.
+   *
+   * If file is not from a leafed partition scheme, result is indeterminate and may error.
+   *
+   * @param name file name
+   * @return
+   */
+  def leaf(name: String): String = name.substring(0, name.indexOf(LeafSeparator))
```

Comment:
Can typenames have '_'s?

Even if that's the case, the typename is completely to the right of the leaf, so that'd be ok?

Reply:
i don't think typenames would be in these file names, they're usually `leaf + type char + uuid`. but yeah, the first _ should delimit the leaf so if there are other ones it should be ok. now, if the last piece of the partition has underscores we're screwed :P

Reply:
Right, the filename doesn't have the typename, so we are fine...

Wait, what's the last piece of partition?  Ok, I'm confused.  Did I find a bug or not:)?

Reply:
well, partitions can technically be anything since it's extensible. i think we have some TODOs around making them be more robust (i.e. we determine the depth of a partition by counting `/`s). we support attribute partitions, so it's possible someone could create a partition scheme like 'yyyy/MM/dd/<attr-foo>' and the encoded attribute value could have underscores in it. it could also have slashes in it, which would also break things. so, yeah it could potentially be a bug but in practice probably not, and we'd need to change some things across the code base to fix it.